# Chapter 2: Variables, Expressions, and Statements

## Section: String Operations

### 🔤 Operators Work Differently with Strings!

Some operators work with strings, but they behave **differently** than with numbers.

We'll explore two operators:
- **`+`** with strings (concatenation)
- **`*`** with strings (repetition)

---
### ➕ The `+` Operator with Strings: Concatenation

**The `+` operator works with strings, but it is NOT addition in the mathematical sense.**

Instead, it performs **concatenation**, which means **joining the strings by linking them end to end**.

Think of it as: **gluing strings together**!

### 📖 Examples from the Book

Let's see the exact examples from the book:

In [1]:
# Example 1: Numbers - Addition
first = 10
second = 15
print(first + second)

25


**Result:** `25` (mathematical addition)

In [2]:
# Example 2: Strings - Concatenation
first = '100'
second = '150'
print(first + second)

100150


**Result:** `100150` (strings joined together, NOT 250!)

**Key Point:** When you use `+` with strings, they're glued together, not added mathematically.

### 🔍 Understanding String Concatenation

In [3]:
# Visual comparison
print("=== With Numbers (Addition) ===")
num1 = 10
num2 = 15
print(f"{num1} + {num2} = {num1 + num2}")

print("\n=== With Strings (Concatenation) ===")
str1 = '10'
str2 = '15'
print(f"'{str1}' + '{str2}' = '{str1 + str2}'")

=== With Numbers (Addition) ===
10 + 15 = 25

=== With Strings (Concatenation) ===
'10' + '15' = '1015'


In [4]:
# More concatenation examples
greeting = "Hello"
name = "World"
message = greeting + name
print(message)

HelloWorld


In [5]:
# Notice: No space between words!
# To add a space, include it in one of the strings:
greeting = "Hello"
name = "World"
message = greeting + " " + name
print(message)

Hello World


In [6]:
# Or include the space in the first string:
greeting = "Hello "
name = "World"
message = greeting + name
print(message)

Hello World


In [7]:
# Concatenating multiple strings
first_name = "John"
middle_name = "Michael"
last_name = "Smith"

full_name = first_name + " " + middle_name + " " + last_name
print(full_name)

John Michael Smith


In [8]:
# Practical example: Building a sentence
subject = "Python"
verb = "is"
adjective = "awesome"

sentence = subject + " " + verb + " " + adjective + "!"
print(sentence)

Python is awesome!


---
### ✖️ The `*` Operator with Strings: Repetition

**The `*` operator also works with strings** by multiplying the content of a string by an integer.

This creates **repetition** - the string is repeated multiple times!

### 📖 Example from the Book

In [9]:
# Book's example
first = 'Test '
second = 3
print(first * second)

Test Test Test 


**Result:** `Test Test Test ` (the string repeated 3 times)

**Note:** The space at the end of 'Test ' is included in each repetition!

### 🔍 Understanding String Repetition

In [10]:
# More repetition examples
print("Ha" * 3)
print("Echo! " * 5)
print("=" * 20)  # Useful for creating separators!

HaHaHa
Echo! Echo! Echo! Echo! Echo! 


In [11]:
# The order doesn't matter
text = "Python "
print(text * 3)      # String first
print(3 * text)      # Number first - same result!

Python Python Python 
Python Python Python 


In [12]:
# Creating patterns
print("*" * 10)
print("Hello")
print("*" * 10)

**********
Hello
**********


In [13]:
# Multiple repetition with spaces
word = "Data"
print(word * 1)      # 1 time
print(word * 2)      # 2 times
print(word * 3)      # 3 times
print((word + " ") * 3)  # With spaces between

Data
DataData
DataDataData
Data Data Data 


In [14]:
# What happens with 0 or negative numbers?
print("Test" * 0)    # Empty string
print("Test" * -1)   # Empty string (negative = 0)

---
### 🎨 Combining Concatenation and Repetition

In [15]:
# You can combine both operations!
border = "=" * 20
title = "Welcome"
message = border + "\n" + title + "\n" + border
print(message)

Welcome


In [16]:
# Creating a fancy box
width = 30
top = "+" + "-" * (width - 2) + "+"
content = "| " + "Python is awesome!" + " " * 8 + "|"
bottom = "+" + "-" * (width - 2) + "+"

print(top)
print(content)
print(bottom)

+----------------------------+
| Python is awesome!        |
+----------------------------+


In [17]:
# Creating a simple ASCII art
line1 = " " * 5 + "*" * 10
line2 = " " * 4 + "*" * 12
line3 = " " * 3 + "*" * 14

print(line1)
print(line2)
print(line3)

     **********
    ************
   **************


---
### ⚠️ Important: Type Matters!

In [20]:
# You CANNOT mix strings and numbers with +
# This will cause an error (uncomment to see):
# result = "Score: " + 100

# Solution 1: Convert number to string
result = "Score: " + str(100)
print(result)

# Solution 2: Use f-strings (modern Python)
score = 100
result = f"Score: {score}"
print(result)

Score: 100
Score: 100


In [ ]:
# String * String doesn't work
# This will cause an error (uncomment to see):
# result = "Hello" * "World"

# String * Number works
result = "Hello" * 3
print(result)

### 🆚 Quick Comparison Table

| Operation | Numbers | Strings | Example |
|-----------|---------|---------|----------|
| **`+`** | Addition | Concatenation (joining) | `"Hello" + "World"` → `"HelloWorld"` |
| **`*`** | Multiplication | Repetition | `"Ha" * 3` → `"HaHaHa"` |
| **`-`** | Subtraction | ❌ Not supported | - |
| **`/`** | Division | ❌ Not supported | - |

---
### 📝 Practice Exercises

In [ ]:
# Exercise 1: Simple concatenation
# Create two variables with your first and last name
# Concatenate them with a space
# Print the result

# Your code here:


In [ ]:
# Exercise 2: Build an email
# Given username = "john" and domain = "example.com"
# Create an email address: john@example.com

username = "john"
domain = "example.com"
# Your code here:


In [ ]:
# Exercise 3: String repetition
# Print "Python" 5 times in a row (no spaces)
# Then print "Python " (with space) 5 times

# Your code here:


In [ ]:
# Exercise 4: Create a separator line
# Create a line of 50 dashes (-)
# Then create a line of 50 equal signs (=)

# Your code here:


In [ ]:
# Exercise 5: Predict the output
# What will these print? Predict before running!

# a)
print("3" + "4")
# Your prediction:

# b)
print("3" * 4)
# Your prediction:

# c)
print(3 + 4)
# Your prediction:

### 🎯 Challenge Exercises

In [ ]:
# Challenge 1: Build a formatted receipt
# Create output like this:
# ==============================
#          RECEIPT
# ==============================
# Item: Coffee
# Price: $5.00
# ==============================

# Your code here:


In [ ]:
# Challenge 2: Create a simple pattern
# Create this pattern:
# *
# **
# ***
# ****
# *****

# Your code here:


In [ ]:
# Challenge 3: Build a URL
# Given:
protocol = "https"
domain = "example.com"
path = "about"
# Create: https://example.com/about

# Your code here:


In [ ]:
# Challenge 4: Combine repetition and concatenation
# Create a box around text:
# ********************
# *   Hello World!   *
# ********************

# Your code here:


### 🔬 Experiments

In [21]:
# Experiment 1: What happens with empty strings?
empty = ""
text = "Hello"

print("Empty + text:", empty + text)
print("Text + empty:", text + empty)
print("Empty * 5:", empty * 5)

Empty + text: Hello
Text + empty: Hello
Empty * 5: 


In [22]:
# Experiment 2: Concatenating many strings
# Multiple ways to do the same thing

# Method 1: Multiple +
result1 = "a" + "b" + "c" + "d"

# Method 2: Building up
result2 = "a"
result2 = result2 + "b"
result2 = result2 + "c"
result2 = result2 + "d"

# Method 3: Using join (advanced)
result3 = "".join(["a", "b", "c", "d"])

print(result1)
print(result2)
print(result3)

abcd
abcd
abcd


In [23]:
# Experiment 3: Mixing quotes
# Single and double quotes can be mixed
text1 = 'Hello' + " " + "World"
text2 = "She said " + "'Hello!'"

print(text1)
print(text2)

Hello World
She said 'Hello!'


---
### 📌 Key Takeaways

- ✅ **`+` with strings** = **Concatenation** (joining strings together)
- ✅ **NOT addition** - `"100" + "150"` gives `"100150"`, not `250`
- ✅ **`*` with strings** = **Repetition** (repeating strings)
- ✅ **`"Test " * 3`** gives `"Test Test Test "`
- ✅ Concatenation links strings **end to end**
- ✅ Remember to include **spaces** if you want them
- ✅ Order doesn't matter for repetition: `"Ha" * 3` = `3 * "Ha"`
- ⚠️ **Cannot mix types**: `"Score: " + 100` causes error
- ✅ Use `str()` to convert numbers to strings
- ✅ String operators work differently than number operators

**Remember:** With strings, `+` glues and `*` repeats!

### 🤔 Reflection Questions

1. What's the difference between `10 + 15` and `"10" + "15"`?
2. What does `"Python" * 4` produce?
3. Why doesn't `"Hello" + 5` work?
4. How can you add a space between concatenated strings?
5. What happens if you multiply a string by 0?

**Your answers:**

1. 
2. 
3. 
4. 
5. 

### 💭 My Notes

*Important points or questions:*

- 
- 
- 